File: frontend/src/api/tasks.ts
import client from './client'

// 任务接口覆盖后台入库/重建的状态、告警、趋势和事件流，用于运维监控页。
export type TaskStatus = 'pending' | 'running' | 'completed' | 'failed'

export interface Task {
  id: string
  knowledge_base_id: string
  task_type?: string
  document_ids: string[]
  status: TaskStatus
  message?: string
  attempt_count?: number
  last_error?: string | null
  started_at?: string | null
  finished_at?: string | null
  locked_at?: string | null
  locked_by?: string | null
  created_at: string
  updated_at: string
}

export interface TaskListResponse {
  items: Task[]
  total: number
}

export interface TaskStatsResponse {
  total: number
  pending: number
  running: number
  completed: number
  failed: number
  retrying: number
  stale_running: number
}

export interface KnowledgeBaseFailureSummary {
  knowledge_base_id: string
  task_count: number
}

export interface TaskOverviewResponse extends TaskStatsResponse {
  active_workers: number
  oldest_pending_age_seconds: number | null
  long_running: number
  recent_failed: number
  recent_retried: number
  knowledge_bases_with_recent_failures: KnowledgeBaseFailureSummary[]
}

export interface TaskAlert {
  code: string
  severity: 'critical' | 'warning'
  message: string
  count: number
  details: Record<string, unknown>
}

export interface TaskAlertListResponse {
  items: TaskAlert[]
  total: number
}

export interface KnowledgeBaseTaskTrend {
  knowledge_base_id: string
  knowledge_base_name: string
  pending: number
  running: number
  recent_failed: number
  previous_failed: number
  failed_delta: number
  recent_retried: number
  previous_retried: number
  retried_delta: number
  recent_completed: number
  previous_completed: number
  completed_delta: number
  updated_at?: string | null
}

export interface KnowledgeBaseTaskTrendListResponse {
  items: KnowledgeBaseTaskTrend[]
  total: number
}

export interface TaskEvent {
  id: string
  task_id: string
  event_type: string
  message: string
  payload?: Record<string, unknown> | null
  created_at: string
}

export interface TaskEventListResponse {
  items: TaskEvent[]
  total: number
}

export const taskApi = {
  get: (id: string): Promise<Task> => client.get(`/tasks/${id}`),
  list: (params?: { knowledge_base_id?: string; status?: TaskStatus; limit?: number }): Promise<TaskListResponse> =>
    client.get('/tasks', { params }),
  // stats 是轻量计数，overview/alerts/trends/events 面向更完整的监控视图。
  stats: (params?: { knowledge_base_id?: string }): Promise<TaskStatsResponse> =>
    client.get('/tasks/stats', { params }),
  overview: (params?: { knowledge_base_id?: string }): Promise<TaskOverviewResponse> =>
    client.get('/tasks/overview', { params }),
  alerts: (params?: { knowledge_base_id?: string }): Promise<TaskAlertListResponse> =>
    client.get('/tasks/alerts', { params }),
  trends: (params?: { limit?: number }): Promise<KnowledgeBaseTaskTrendListResponse> =>
    client.get('/tasks/trends', { params }),
  events: (id: string, params?: { limit?: number }): Promise<TaskEventListResponse> =>
    client.get(`/tasks/${id}/events`, { params }),
}

File: backend/app/api/endpoints/tasks.py
"""后台任务观测接口。

这些接口服务于任务观测页和文档页告警：列表看当前任务，overview/alerts/trends 看 worker
是否积压、卡住或近期失败增多。
"""

from fastapi import APIRouter, Depends, Query

from ..deps import get_task_service
from ...schemas.task import (
    IngestTaskListResponse,
    IngestTaskResponse,
    KnowledgeBaseTaskTrendListResponse,
    TaskAlertListResponse,
    TaskEventListResponse,
    TaskOverviewResponse,
    TaskStatsResponse,
)
from ...services.ingest_service import TaskService

router = APIRouter(prefix="/tasks")


@router.get("", response_model=IngestTaskListResponse)
async def list_tasks(
    knowledge_base_id: str | None = None,
    status: str | None = None,
    limit: int = Query(default=50, ge=1, le=200),
    service: TaskService = Depends(get_task_service),
) -> IngestTaskListResponse:
    payload = service.list_tasks(
        knowledge_base_id=knowledge_base_id,
        status=status,
        limit=limit,
    )
    return IngestTaskListResponse(**payload)


@router.get("/stats", response_model=TaskStatsResponse)
async def get_task_stats(
    knowledge_base_id: str | None = None,
    service: TaskService = Depends(get_task_service),
) -> TaskStatsResponse:
    payload = service.get_task_stats(knowledge_base_id=knowledge_base_id)
    return TaskStatsResponse(**payload)


@router.get("/overview", response_model=TaskOverviewResponse)
async def get_task_overview(
    knowledge_base_id: str | None = None,
    service: TaskService = Depends(get_task_service),
) -> TaskOverviewResponse:
    """返回任务总览，包含 stale running、active workers 和近期失败等监控指标。"""
    payload = service.get_task_overview(knowledge_base_id=knowledge_base_id)
    return TaskOverviewResponse(**payload)


@router.get("/alerts", response_model=TaskAlertListResponse)
async def get_task_alerts(
    knowledge_base_id: str | None = None,
    service: TaskService = Depends(get_task_service),
) -> TaskAlertListResponse:
    """把总览指标转换成前端可直接展示的告警项。"""
    payload = service.get_task_alerts(knowledge_base_id=knowledge_base_id)
    return TaskAlertListResponse(**payload)


@router.get("/trends", response_model=KnowledgeBaseTaskTrendListResponse)
async def get_task_trends(
    limit: int = Query(default=20, ge=1, le=100),
    service: TaskService = Depends(get_task_service),
) -> KnowledgeBaseTaskTrendListResponse:
    payload = service.get_task_trends(limit=limit)
    return KnowledgeBaseTaskTrendListResponse(**payload)


@router.get("/{task_id}", response_model=IngestTaskResponse)
async def get_task(
    task_id: str,
    service: TaskService = Depends(get_task_service),
) -> IngestTaskResponse:
    record = service.get_task(task_id)
    return IngestTaskResponse(**record)


@router.get("/{task_id}/events", response_model=TaskEventListResponse)
async def list_task_events(
    task_id: str,
    limit: int = Query(default=100, ge=1, le=500),
    service: TaskService = Depends(get_task_service),
) -> TaskEventListResponse:
    """返回单任务事件时间线，用于定位任务卡在哪个阶段。"""
    payload = service.get_task_events(task_id, limit=limit)
    return TaskEventListResponse(**payload)
